javaslang/src/main/java/javaslang/collection/AbstractMap.java

```diff
@@ -47,6 +47,31 @@ public M put(Tuple2<? extends K, ? extends V> entry) {
 
     @SuppressWarnings("unchecked")
     @Override
+    public M putWith(K key, V value, BiFunction<? super V, ? super V, ? extends V> merge) {
+        Objects.requireNonNull(merge, "the merge function is null");
+        Option<V> currentValue = get(key);
```

Comment:
One hint (I know it is nasty): Please use the `final` keyword on _local vars_ if possible. That aligns the code style to the  Javaslang code base.

Once I decided to do it from the ground up while building Javaslang because Scala has `var` and `val`. One of the most evil enemies of functional programming is Mr. Mutation. Having final vars hardens our code.

The only thing I don't like about `final` is its verbosity. In an utopic Java language it should have been the default. But with [JEP #286](http://openjdk.java.net/jeps/286) comes the rescue (I pray every night for it). Then all of our local vars will be either `var` or `val`, of course most of them `val`.